code/modules/research/ordnance/scipaper_publisher.dm

```diff
@@ -0,0 +1,220 @@
+			data["purchaseableBoosts"][partner.type] = list()
+			for(var/node_id in linked_techweb.get_available_nodes())
+				// Not from our partner
+				if(!(node_id in partner.boosted_nodes))
+					continue
+				if(!partner.allowed_to_boost(linked_techweb, node_id))
+					continue
+				data["purchaseableBoosts"][partner.type] += node_id
+	return data
+
+/datum/computer_file/program/scipaper_program/ui_act(action, params)
+	. = ..()
+	if (.)
+		return
+
+	switch(action)
+		if("et_alia")
+			paper_to_be.et_alia = !paper_to_be.et_alia
+		// Handle the publication
+		if("publish")
+			publish()
+		// For every change in the input, we correspond it with the paper_data list and update it.
+		if("rewrite")
+			if(length(params))
+				for (var/changed_entry in params)
+					if (changed_entry == "title")
+						paper_to_be.title = sanitize(params[changed_entry])
+					if (changed_entry == "author")
+						paper_to_be.author = sanitize(params[changed_entry])
+					if (changed_entry == "abstract")
+						paper_to_be.abstract = sanitize(params[changed_entry])
+		if("change_tab")
+			current_tab = params["new_tab"]
+		if("select_file") // Selecting new file will necessitate a change in paper type. This will be done on select_experiment and not here.
+			if(selected_file)
+				UnregisterSignal(selected_file, COMSIG_MODULAR_COMPUTER_FILE_DELETED)
+			paper_to_be.set_experiment() // Clears the paper info.
+			for(var/datum/computer_file/data/ordnance/ordnance_data in holder.stored_files)
+				if(ordnance_data.uid == params["selected_uid"])
+					selected_file = ordnance_data
+					RegisterSignal(selected_file, COMSIG_MODULAR_COMPUTER_FILE_DELETED, .proc/recheck_file_presence)
+					return
+		if("select_experiment")
+			var/ex_path = text2path(params["selected_expath"])
+			var/variable = selected_file.possible_experiments[text2path(params["selected_expath"])]
+			var/data = null
+			if(ispath(ex_path, /datum/experiment/ordnance/explosive))
+				paper_to_be = paper_to_be.clone_into(/datum/scientific_paper/explosive)
+			if(ispath(ex_path, /datum/experiment/ordnance/gaseous))
+				paper_to_be = paper_to_be.clone_into(/datum/scientific_paper/gaseous)
+			data = selected_file.return_data()
+			paper_to_be.set_experiment(ex_path, variable, data)
+		if("select_tier")
+			paper_to_be.set_tier(params["selected_tier"])
+		if("select_partner")
+			paper_to_be.set_partner(text2path(params["selected_partner"]), linked_techweb)
+		if("purchase_boost")
+			var/datum/scientific_partner/partner = locate(text2path(params["boost_seller"])) in SSresearch.scientific_partners
+			var/datum/techweb_node/node = SSresearch.techweb_node_by_id(params["purchased_boost"])
+			if(partner && node)
+				if(partner.purchase_boost(linked_techweb, node))
+					computer.say("Purchase succesful.")
+					playsound(computer, 'sound/machines/ping.ogg', 25)
+					return TRUE
+			playsound(computer, 'sound/machines/terminal_error.ogg', 25)
+	SStgui.update_uis(src)
```

Comment:
doesn't returning true do this too? why do it manually

Reply:
I didn't know it does. I'll get rid of it.